### ISSUE TITLE
Unable to write in a BQ table, with the new spark connector update - issue persists

### ISSUE THREAD

**Issue by katerina-kogan** [GoogleCloudDataproc/spark-bigquery-connector#1149]
Initially discussed here: https://github.com/GoogleCloudDataproc/spark-bigquery-connector/issues/658
Seems that the issue still persists.
Given:
1. dbt python model (dbt cloud) called fct_model
2.Simple schema:
root
 |-- customer_id: integer (nullable = true)
 |-- created: timestamp (nullable = true)
 |-- attempt_number: integer (nullable = false)
 |-- status_desc: string (nullable = true)
 |-- payment_type: string (nullable = true)
3.  The model runs on Dataproc Serverless with the built-in bigquery connector. It runs successfully only the first time, creates the fct_model table in bigquery and populates it with data. 
4. Every subsequent model rerun gives an error:
" Error in custom provider, java.lang.IllegalArgumentException: com.google.cloud.bigquery.connector.common.BigQueryConnectorException$InvalidSchemaException: Destination table’s schema is not compatible with dataframe’s schema
" 
even if it runs over exactly same data.

What I have noticed is that the issue occurs only when we have defined the model columns definition in DBT yml file with persist_doc config (https://docs.getdbt.com/reference/resource-configs/persist_docs):
- name: fct_model
    description: fct_model
    config:
        submission_method: cluster
    columns:
      - name: customer_id
        description: customer_id
      - name: created
        description: created
      - name: attempt_number
        description: attempt_number
      - name: status_desc
        description: status_desc
      - name: payment_type
        description: payment_type

This definition will update column description in BigQuery fct_model table automatically.
When we dont have the columns defined as above, we can successfully rerun the model as many times as we want:
- name: fct_model
    description: fct_model
    config:
        submission_method: cluster

Also, issue does not occur with older bigquery connector (0.22).
Seems like we are missing just some config/parameter here... Would appreciate an assistance with the issue,  thank you!

**Comment by vishalkarve15** [comment]
@katerina-kogan Can you please share the exact steps and a code sample to reproduce the issue?

**Comment by katerina-kogan** [comment]
@vishalkarve15 , please, kindly check the steps:
1. dbt cloud python model in `fct_model.py`:

```
import pyspark.sql.functions as F
from pyspark.sql.window import Window

TWO_YEARS_DAYS = 365*2

import pyspark.sql.functions as F
from pyspark.sql.window import Window

TWO_YEARS_DAYS = 365*2

def model(dbt, session):
    session.conf.set("viewsEnabled","true")

    table1_df = dbt.source("dataset1", "table1")
    table2_df = dbt.source("dataset1", "table2")
    table3_df = dbt.source("dataset2", "table3")
    table4_df = dbt.source("dataset2", "table4")
    table5_df = dbt.ref("ref_table")


   return table1_df \
      .join(table2_df, table1_df.o_id==table2_df.oo_id,"inner") \
      .join(table4_df, table4_df.o_id==table2_df.id,"inner") \
      .join(table3_df, table4_df.g_id==table3_df.id,"inner") \
      .join(table5_df, table4_df.status==table5_df.status_code,"left") \
      .filter(table1_df.created >= F.date_sub(F.current_date(), TWO_YEARS_DAYS)) \
      .filter(table4_df.created >= F.date_sub(F.current_date(), TWO_YEARS_DAYS)) \
      .withColumn("attempt_number", F.row_number().over(Window.partitionBy(table1_df.id).orderBy(table2_df.created_at))) \
      .withColumn("payment_type", F.coalesce(F.get_json_object(table4_df["extra"], "$.result.data.payment_type"), table3_df["payment_type"])) \
      .select(
          table2_df["customer_id"], 
          F.to_date(table1_df["created"]).alias("created"),
          F.col("attempt_number"),
          F.col("status_desc").alias("status"),
          F.col("payment_type")
          ).repartition(10)
```

2. Description of the model in `model.yml` :
```
version: 2

models:
- name: fct_model
    description: fct_model
    config:
        submission_method: cluster
    columns:
      - name: customer_id
        description: customer_id
      - name: created
        description: created
      - name: attempt_number
        description: attempt_number
      - name: status_desc
        description: status_desc
      - name: payment_type
        description: payment_type
```
3. The model runs on Dataproc Serverless via "dbt built" command in dbt cloud
4. The first run is successful, it creates a `fct_model ` table  in bigquery and adds data
5. Any subsequent run results in error: "Destination schema is not compatible"

P.S. as stated in the issue description, if `columns` section is not there in `model.yml` file above, the model runs fine. But once the section is added, model starts subsequently failing. Cant say it is 100% relevance, maybe unfortunate coincidence with something wrong going behind the scene

Thank you!

**Comment by katerina-kogan** [comment]
updated issue details: according to the [doc ](https://github.com/GoogleCloudDataproc/spark-bigquery-connector) the Dataproc serverless uses built in connector. 
I also tested on a cluster and specified connector version directly as doc suggested:
SPARK_BQ_CONNECTOR_VERSION=0.34.0
and it works fine

**Comment by vishalkarve15** [comment]
It works fine, you mean, this issue does not occur in the 0.34.0 connector? Even including the `columns` section ?

**Comment by katerina-kogan** [comment]
I believe the issue is with built in connector, but the doc doesnt say which exact version it is

**Comment by vishalkarve15** [comment]
If you're using Serverless 2.1, it comes with built in 0.28.1. See https://cloud.google.com/dataproc-serverless/docs/concepts/versions/spark-runtime-versions

**Comment by katerina-kogan** [comment]
Hi @vishalkarve15 , we have developed on Dataproc Cluster, Spark 3.3.2
https://cloud.google.com/dataproc/docs/concepts/versioning/dataproc-release-2.1

**Comment by vishalkarve15** [comment]
That one comes with 0.27.1. We have updated the documentation to reflect this yesterday.
I'm closing this since it has been fixed in the latest 0.34.0.
Feel free to reopen if you still face issues.

**Comment by katerina-kogan** [comment]
Sorry folks, have to open another issue: https://github.com/GoogleCloudDataproc/spark-bigquery-connector/issues/1158
Same error, but now happening in 0.35 version
